### ISSUE TITLE
Syntaxhighlighting quirks

### ISSUE THREAD

**Issue by Skillmon** [lervag/vimtex#1861]
There are a few issues with the syntax highlighting.

- `expl3` syntax doesn't stack, `\ExplSyntaxOn` is just a no-op after `\ExplSyntaxOn` until the end of the `expl3`-syntax. So after `\ExplSyntaxOn\ExplSyntaxOn\ExplSyntaxOff` the `expl3` is off, the syntax highlighting will currently continue to use the rules of `expl3` instead.

- In TeX, only 9 parameters can be used, so the regular expression used for parameters shouldn't be `#\d\+`. What instead might be used doubled in TeX is the `#`. This is necessary if one defines a macro in the definition of another, so the following would be valid TeX-code:

    ```tex
    \def\foo#1{\def\bar##1{#1##1#1}}
    ```

    where `#1` would refer to the parameter of `\foo`, and `##1` to the parameter of `\bar`. So a better regular expression for parameters would, imho, be `#\+\d`. This isn't limited to the `expl3`-syntax, but to all of the syntax highlighting rules, afaics.

**Comment by Rmano** [comment]
>     * `expl3` syntax doesn't stack, `\ExplSyntaxOn` is just a no-op after `\ExplSyntaxOn` until the end of the `expl3`-syntax. So after `\ExplSyntaxOn\ExplSyntaxOn\ExplSyntaxOff` the `expl3` is off, the syntax highlighting will currently continue to use the rules of `expl3` instead.

This is probably my fault, because I have sent to @lervag  this piece of code:

```latex
\ExplSyntaxOn
    \str_new:N \g_myfile_name_str
    \str_gset:Nn \g_myfile_name_str {myfile.txt}
    \str_new:N \g_myfile_old_mdfive_str
    \str_new:N \l_myfile_mdfive_str
    \cs_new_protected:Npn \getmdfive % get hash is not expandable, so protect
    {
        \file_get_mdfive_hash:nN {\str_use:N \g_myfile_name_str} \l_myfile_mdfive_str
        \str_use:N \l_myfile_mdfive_str
    }
    \cs_new:Npn \getoldmdfive
    {
        \str_use:N \g_myfile_old_mdfive_str
    }
    \cs_new:Npn \IfMyfileChanged #1 #2
    {
        \str_if_eq:NNTF \g_myfile_old_mdfive_str \l_myfile_mdfive_str {#2} {#1}
    }
    \AtEndDocument
    {
        \iow_now:cx { @auxout }
        {
            \token_to_str:N \ExplSyntaxOn
            \iow_newline:
            \str_gset:Nn \token_to_str:N \g_myfile_old_mdfive_str {\str_use:N \l_myfile_mdfive_str}
            \iow_newline:
            \token_to_str:N \ExplSyntaxOff
        }
    }
\ExplSyntaxOff
```
... seems a difficult thing to separate...

**Comment by lervag** [comment]
I'm thinking about a possible solution: Do not allow nested `\ExplSyntaxOn|Off` regions at "top level", only inside "argument groups" or "TeX groups". E.g.:

```tex
% Example 1
\ExplSyntaxOn    % expl3 region activate
...              % expl3 region on
\ExplSyntaxOn    % expl3 region on
\ExplSyntaxOff   % expl3 region disable
...              % expl3 region off
\ExplSyntaxOff   % expl3 region off

% Example 2
\ExplSyntaxOn    % expl3 region activate
...              % expl3 region on
{\ExplSyntaxOn   % expl3 region activete 2
...              % expl3 region on nested
\ExplSyntaxOff}  % expl3 region disable 2
...              % expl3 region off
\ExplSyntaxOff   % expl3 region off
```

@Skillmon Regarding parameters like `#\+\d`, would you "accept" a simple version of this that does not really recognize these nested parameters except color them "properly"? I'm pushing that now, so you can test what I mean.

**Comment by Skillmon** [comment]
I'm not currently on the PC on which I test vimtex (sorry, not a regular user just yet, but only because of laziness, it's hard to migrate). But I took a quick look at the file and it looks just like what I would've done, so I guess I'm ok with this (detecting those nested `\def`s is futile, again, TeX is arbitrarily hard to parse right without actually using TeX.

**Comment by lervag** [comment]
> I'm not currently on the PC on which I test vimtex (sorry, not a regular user just yet, but only because of laziness, it's hard to migrate).

No worries, no need to apologize! :)

> But I took a quick look at the file and it looks just like what I would've done, so I guess I'm ok with this (detecting those nested `\def`s is futile, again, TeX is arbitrarily hard to parse right without actually using TeX.

Exactly. TeX is very difficult to parse, so I've tried to make the syntax rules as simple as possible while still being relevant and useful. I do not aim to have "full support".
